src/Microsoft.DotNet.Interactive.Tests/LanguageKernelPackageTests.cs

```diff
@@ -1101,40 +1101,90 @@ public async Task Pound_r_nuget_works_immediately_after_a_language_selector(Lang
         [InlineData(Language.FSharp)]
-        public async Task Pound_r_nuget_should_display_only_requested_packages(Language defaultLanguageKernel)
+        public async Task Pound_r_nuget_should_display_only_requested_packages_for_first_submission(Language defaultLanguageKernel)
         {
             var kernel = CreateCompositeKernel(defaultLanguageKernel);
 
             var code = @"
-#r ""nuget:Google.Protobuf""
 #r ""nuget:Microsoft.ML.OnnxTransformer,1.4.0""
 ";
 
             var command = new SubmitCode(code);
-
             var result = await kernel.SendAsync(command);
-
             using var events = result.KernelEvents.ToSubscribedList();
 
             using var _ = new AssertionScope();
 
             var expectedDisplayed = new[]
             {
-                "Google.Protobuf, 3.17.3",
                 "Microsoft.ML.OnnxTransformer, 1.4.0"
             };
 
             var notExpectedDisplayed = new[]
             {
-                "Microsoft.ML.KMeansClustering",
-                "Microsoft.ML.PCA",
-                "Microsoft.ML.StandardTrainers",
-                "Microsoft.ML.Transforms",
-                "Microsoft.ML.CpuMath",
-                "Microsoft.ML.DataView",
-                "Microsoft.ML.OnnxRuntime",
-                "Newtonsoft.Json",
-                "System.CodeDom"
+                "google.protobuf, 3.5.1",
+                "microsoft.ml.kmeansclustering",
+                "microsoft.ml.pca",
+                "microsoft.ml.standardtrainers",
+                "microsoft.ml.transforms",
+                "microsoft.ml.cpumath",
+                "microsoft.ml.dataview",
+                "microsoft.ml.onnxruntime",
+                "newtonsoft.json",
+                "system.codedom"
+            };
+
+            events.OfType<DisplayedValueUpdated>()
+                  .Where(v => v.Value is InstallPackagesMessage)
+                  .Last().Value
+                  .As<InstallPackagesMessage>()
+                  .InstalledPackages
+                  .Aggregate((s, acc) => acc + " & " + s)
+                  .Should()
+                  .ContainAll(expectedDisplayed)
+                  .And
+                  .NotContainAny(notExpectedDisplayed);
+        }
+
+        [Theory]
+        [InlineData(Language.CSharp)]
+        [InlineData(Language.FSharp)]
+        public async Task Pound_r_nuget_should_display_only_requested_packages_for_subsequent_submission(Language defaultLanguageKernel)
+        {
+            var kernel = CreateCompositeKernel(defaultLanguageKernel);
+
+            var codeFirstSubmission = @"
+#r ""nuget:Microsoft.ML.OnnxTransformer,1.4.0""
+";
+            var command1 = new SubmitCode(codeFirstSubmission);
+            var _result1 = await kernel.SendAsync(command1);
+
+            var codeSecondSubmission = @"
+#r ""nuget:Google.Protobuf, 3.5.1""
+";
+            var command2 = new SubmitCode(codeSecondSubmission);
+            var result2 = await kernel.SendAsync(command2);
```

Comment:
I would subscribe to the event before submitting the command that is to be tested

Reply:
It's the third submission I want events for, and that is updated, per your comment.